### ISSUE TITLE
Tensorflow consumes much more memory than expected

### ISSUE THREAD

**Issue by fesun** [tensorflow/tensorflow#9823]
My model has four CPU variables:
[500M, 3] tf.int32
[500M] tf.float32
[500M] tf.float32 (FTRL accumulate slot)
[500M] tf.float32 (FTRL linear slot)
expected memory consumption should be (500M * 6) * 4 = 12G, however tensorflow used 20G memory.

When I increased 500M to 1B, total memory usage is 40G, seems tensorflow do allocate much more memory than needed, any idea? By the way I am not using any tcmalloc stuff.
I also used timeline show_memory to print allocated tensor size, everything is consistent with my calculating.

**Comment by fesun** [comment]
I traced it using strace and found two mmap called by different two threads, so actually memory consumption is 2x than needed.

Here is the code:
```python
import tensorflow as tf
import time

bucket_size = 1000000000

a = tf.get_variable("W", [bucket_size], initializer=tf.zeros_initializer, dtype=tf.float32, trainable=False)

with tf.Session() as sess:
    print("start to initialize variable")
    time.sleep(10)
    sess.run(a.initializer)
    print("initialize variable done")

time.sleep(10)
```

Here is the strace output
```
[pid 25768] write(1, "start to initialize variable\n", 29start to initialize variable
 <unfinished ...>
[pid 25817] futex(0x26c92e4, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
[pid 25768] <... write resumed> )       = 29
[pid 25768] select(0, NULL, NULL, NULL, {10, 0}) = 0 (Timeout)
[pid 25768] mmap(NULL, 4294967296, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7f2ec8000000
[pid 25768] futex(0x26c92e4, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x26c92e0, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
[pid 25817] <... futex resumed> )       = 0
[pid 25768] futex(0x7ffdfa482a9c, FUTEX_WAIT_PRIVATE, 1, NULL <unfinished ...>
[pid 25817] futex(0x26c92b8, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25817] futex(0x26c9264, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x26c9260, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
[pid 25816] <... futex resumed> )       = 0
[pid 25817] mmap(NULL, 2097152, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0 <unfinished ...>
[pid 25816] futex(0x26c9238, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 25817] <... mmap resumed> )        = 0x7f307f9ff000
[pid 25816] <... futex resumed> )       = 0
[pid 25817] munmap(0x7f307f9ff000, 2097152 <unfinished ...>
[pid 25816] futex(0x26c9264, FUTEX_WAIT_PRIVATE, 3, NULL <unfinished ...>
[pid 25817] <... munmap resumed> )      = 0
[pid 25817] mmap(NULL, 4190208, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7f307f800000
[pid 25817] munmap(0x7f307fa00000, 2093056) = 0
[pid 25817] mmap(NULL, 4294967296, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS|MAP_NORESERVE, -1, 0) = 0x7f2dc8000000
[pid 25817] futex(0x26c9264, FUTEX_WAKE_OP_PRIVATE, 1, 1, 0x26c9260, {FUTEX_OP_SET, 0, FUTEX_OP_CMP_GT, 1}) = 1
[pid 25816] <... futex resumed> )       = 0
[pid 25817] futex(0x7ffdfa482a9c, FUTEX_CMP_REQUEUE_PRIVATE, 1, 2147483647, 0x7ffdfa482a70, 2) = 1
[pid 25816] futex(0x26c9238, FUTEX_WAKE_PRIVATE, 1 <unfinished ...>
[pid 25817] futex(0x26c92e4, FUTEX_WAIT_PRIVATE, 3, NULL <unfinished ...>
[pid 25816] <... futex resumed> )       = 0
[pid 25768] <... futex resumed> )       = 0
[pid 25816] futex(0x26c9264, FUTEX_WAIT_PRIVATE, 5, NULL <unfinished ...>
[pid 25768] futex(0x7ffdfa482a70, FUTEX_WAKE_PRIVATE, 1) = 0
[pid 25768] write(1, "initialize variable done\n", 25initialize variable done
```

Both 25768 25817 called mmap to allocate 4G memory. Why?

**Comment by fesun** [comment]
I seems that this code actually initialized two variables, one is zero const with 4G, the other is W variable, then assign zero const to W. I will check the python code to see what I can do to avoid it.

**Comment by fesun** [comment]
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/variables.py#L276 If initializer is provided and callable, how about passing variable to initializer instead of initialize and then assign it?
Codebase now has one zero initializer operator to initialize variables, https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/kernels/zero_initializer_op.h, we need to add more.
Any idea?

**Comment by yaroslavvb** [comment]
You can already do this, but, I'm not sure that will avoid a copy. There's this ZeroInitialize to adress this use-case -- https://github.com/tensorflow/tensorflow/commit/f38886ba307314cc0ae5b15279f5693f39c92f21

**Comment by yaroslavvb** [comment]
BTW, easier way than strace is to look at tensor allocation/deallocation messages with https://github.com/yaroslavvb/memory_util

**Comment by girving** [comment]
@benoitsteiner Would it be feasible to add a graph rewriting optimization to do in-place assignment for simple cases?

**Comment by benoitsteiner** [comment]
It's on our to-do list, we just lacked a good reason to prioritize this ahead of other rewrites.

**Comment by fesun** [comment]
@benoitsteiner  I am training super sparse model, no memory left due to this 2x memory issue. This is very general case, I don't want to spend effort to fix it for my own, and  several weeks after my own fix you guys fix and commit it, what if I make the change and send pull request?

**Comment by girving** [comment]
Cc @rmlarsen too, since you're name is on comments in `AssignOp` that buffer forwarding rarely triggers.  Looks like @fesun is offering to submit a PR; can we direct him to the right change to make?

**Comment by fesun** [comment]
@rmlarsen ping...

**Comment by fesun** [comment]
@girving @rmlarsen @benoitsteiner  Could you guys make some decision? It's our blocking issue, long time no response.

**Comment by rmlarsen** [comment]
Sorry for not noticing this thread earlier. Let me take a look.

**Comment by fesun** [comment]
@rmlarsen any update? If you guys don't have resource now, I can make the change, but we need to talk and agree the design.

**Comment by fesun** [comment]
Since googler didn't response for very long time, I made one minimal change on my codebase to solve this, and post my solution here in case others are also confused by this:
1) Tensorflow provides ones_initializer/zeros_initializer for initializing variables to 1/0, what tensorflow actually does is creating one const tensor and then assign it to variable, however constant operator holds tensor as its instance member, so its memory will not be freed unless the operator instance is deleted. Tensorflow AssignOp intent to try it's best to avoid memory copy(https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/assign_op.h#L47), this happens only when input tensor refcount is 1, while const tensor refcount is always greater than 1.
**Solution: write custom operator to create normal 0/1 tensor**
2) As AssignOp comments said, conservative constraints make buffer forwarding unlikely to happen very often, my model doesn't leverage GPU and any other advanced communication tech, so I set both gpu and nic compatible to false.
3) patch desired optimizer to use custom 0/1 initializer(slot creation)

**Comment by rmlarsen** [comment]
@fesun Good job figuring this out. My apologies for not getting back to you. I think we can probably implement a more general solution by doing a graph analysis and strip off the buffer constraints when they are not needed, or perhaps propagate them to dependencies, so the bufferes are compatible, perhaps similar to 
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/executor.cc#L718

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activityand the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tatatodd** [comment]
@rmlarsen any updates on the more general solution?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Awaiting TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging TensorFlower: It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by TylerBalsam** [comment]
@fesun Do you happen to have a github link handy to that custom operator?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatianashp: It has been 30 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
@rmlarsen Was the general solution that you mentioned ever implemented?

**Comment by rmlarsen** [comment]
@tatianashp Yes, this was implemented about a month ago:

The graph pass is here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/grappler/optimizers/memory_optimizer.cc#L1226

And the code using the result here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/assign_op.h#L58

**Comment by tatianashp** [comment]
Great. Thank you!

I am closing the issue.
